File: src/main/java/com/parmet/squashlambdas/match/OtherPlayersParser.java
package com.parmet.squashlambdas.match;

import com.google.common.collect.ImmutableSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

class OtherPlayersParser {
  private static final Logger log = LogManager.getLogger();

  private static final Pattern HAS_JOINED =
      Pattern.compile(".*Parmet, (.*) has joined your reservation.*");

  private static final Pattern OTHER_PLAYERS =
      Pattern.compile(".* Other players in the slot: (.*) Court:.*");

  // Name can be followed by an email in parentheses.
  private static final Pattern WITH =
      Pattern.compile(".* With: (.*) (\\(.*\\))? Court:.*");

  public static ImmutableSet<String> parse(String body) {
    Matcher hasJoined = HAS_JOINED.matcher(body);
    if (hasJoined.matches()) {
      return ImmutableSet.of(hasJoined.group(1));
    } else {
      Matcher otherPlayers = OTHER_PLAYERS.matcher(body);
      if (otherPlayers.matches()) {
        return ImmutableSet.of(otherPlayers.group(1));
      } else {
        Matcher with = WITH.matcher(body);
        if (with.matches()) {
          return ImmutableSet.of(with.group(1));
        } else {
          log.info("Did not find other players in body {}", body);
          return ImmutableSet.of();
        }
      }
    }
  }
}

File: src/main/kotlin/com/parmet/squashlambdas/activity/OtherPlayersParser.kt
package com.parmet.squashlambdas.activity

import mu.KotlinLogging
import java.util.regex.Pattern

internal object OtherPlayersParser {
    private val logger = KotlinLogging.logger { }

    private val HAS_JOINED =
        Pattern.compile(".*Parmet, (.*) has joined your reservation.*")

    private val OTHER_PLAYERS =
        Pattern.compile(".* Other players in the slot: (.*) Court:.*")

    // Name can be followed by an email in parentheses.
    private val WITH =
        Pattern.compile(".* With: (.*) (\\(.*\\))? Court:.*")

    fun parse(body: String): Set<String> {
        val hasJoined = HAS_JOINED.matcher(body)
        if (hasJoined.matches()) {
            return setOf(hasJoined.group(1))
        } else {
            val otherPlayers = OTHER_PLAYERS.matcher(body)
            return if (otherPlayers.matches()) {
                setOf(otherPlayers.group(1))
            } else {
                val with = WITH.matcher(body)
                if (with.matches()) {
                    setOf(with.group(1))
                } else {
                    logger.debug { "Did not find other players in body $body" }
                    setOf()
                }
            }
        }
    }
}
